hledger-lib/Hledger/Data/Amount.hs

```diff
@@ -759,22 +765,27 @@ sumSimilarAmountsUsingFirstPrice a b = (a + b){aprice=p}
 
 -- | Filter a mixed amount's component amounts by a predicate.
 filterMixedAmount :: (Amount -> Bool) -> MixedAmount -> MixedAmount
-filterMixedAmount p (Mixed as) = Mixed $ filter p as
+filterMixedAmount p (Mixed ma) = Mixed $ M.filter p ma
 
 -- | Return an unnormalised MixedAmount containing exactly one Amount
 -- with the specified commodity and the quantity of that commodity
 -- found in the original. NB if Amount's quantity is zero it will be
 -- discarded next time the MixedAmount gets normalised.
 filterMixedAmountByCommodity :: CommoditySymbol -> MixedAmount -> MixedAmount
-filterMixedAmountByCommodity c (Mixed as) = Mixed as'
-  where
-    as' = case filter ((==c) . acommodity) as of
-            []   -> [nullamt{acommodity=c}]
-            as'' -> [sum as'']
+filterMixedAmountByCommodity c (Mixed ma)
+  | M.null ma' = mixedAmount nullamt{acommodity=c}
+  | otherwise  = Mixed ma'
+  where ma' = M.filterWithKey (\k _ -> (==c) $ fst k) ma
 
 -- | Apply a transform to a mixed amount's component 'Amount's.
 mapMixedAmount :: (Amount -> Amount) -> MixedAmount -> MixedAmount
-mapMixedAmount f (Mixed as) = Mixed $ map f as
+mapMixedAmount f (Mixed ma) = mixed . map f $ toList ma
+
+-- | Apply a transform to a mixed amount's component 'Amount's, which does not
+-- affect the key of the amount (i.e. doesn't change the commodity, price
+-- commodity, or unit price amount). This condition is not checked.
```

Comment:
Why provide/use this.. is it faster ? Easier to use ? I guess it's unexported/for internal use only ?

Reply:
Internal use only. Any function which does not change the key used can just operate on the values, preventing the need to deconstruct and reconstruct the Map each time. Many commonly used functions meet this criteria, so it will save a lot of heap space.